File: Trapping Rain Water.cpp
class Solution {
public:
    int trap(int A[], int n) {
        // Start typing your C/C++ solution below
        // DO NOT write int main() function
        if (n == 0) return 0;
        
        int wall = 0, water = 0, cur = 0;
        
        for (int i = 0; i < n; i++) {
            if (A[i] > A[wall]) {
                wall = i;
            }
        }
        
        for (int i = 0; i < wall; i++) {
            if (cur > A[i]) {
                water += cur - A[i];
            } else {
                cur = A[i];
            }
        }
        
        cur = 0;
        for (int i = n-1; i > wall; i--) {
            if (cur > A[i]) {
                water += cur - A[i];
            } else {
                cur = A[i];
            }
        }
        
        return water;
        
    }
};

File: Trapping Rain Water.cs
public class Solution {
    public int Trap(int[] height) {
        var result = 0;        
        var leftMax = 0;
        var rightMax = 0;
        var i = 0;
        var j = height.Length - 1;
        
        // The water trapped at any element = min( max_left, max_right) – arr[i]
        while (i <= j) {
            if (height[i] < height[j]) {
                if (height[i] > leftMax) {
                    leftMax = height[i];
                } else {
                    result += leftMax - height[i];
                }
                i++;
            } else {
                if (height[j] > rightMax) {
                    rightMax = height[j];
                } else {
                    result += rightMax - height[j];
                }
                j--;
            }
        }
        
        return result;
    }
}
